cli/client.go

```diff
@@ -403,14 +404,58 @@ func VersionAction(c *cli.Context) error {
 	return nil
 }
 
-func checkBaseURL(c *cli.Context) (*url.URL, []rpc.DialOption, error) {
+var defaultBaseURL = "https://app.viam.com:443"
+
+func checkBaseURL(c *cli.Context, auth *config) (*url.URL, []rpc.DialOption, error) {
+	// If base URL was not specified, assume cached base URL. If no base URL is
+	// cached, assume default base URL.
 	baseURL := c.String(baseURLFlag)
+	if baseURL == "" {
+		baseURL = defaultBaseURL
+		if auth.Auth != nil && auth.Auth.BaseURL != "" {
+			baseURL = auth.Auth.BaseURL
+		}
+	}
+
+	if auth.Auth != nil && auth.Auth.BaseURL != baseURL {
+		return nil, nil, fmt.Errorf("cached base URL for this session is %q, "+
+			"please logout and login again to use provided base URL %q", auth.Auth.BaseURL, baseURL)
```

Comment:
So helpful, thanks for adding this!

Out of curiosity, why is it important to have the user log out? Could we just log them out of the old session next time they run `viam auth` with a new baseurl?

Reply:
> Could we just log them out of the old session next time they run viam auth with a new baseurl?

We could! `viam logout` basically just clears the auth cache. I don't love the idea of silently clearing the cache when the base URL changes for `viam login`. The caching behavior is already a little confusing here, and I kind of like having the user explicitly "clear" their cache by logging out and logging back in with their desired base URL after they get this error. I don't feel too strongly though, if you feel that would be more ergonomic.

Reply:
That makes sense, thanks for explaining. Personally the concept of the cache was new to me before this PR/ticket. I had actually thought that the default URL was always app.viam.com, and that if you want to connect/send requests to a different deployment you had to include that URL in every request. And I had thought the CLI could have you logged into multiple sessions at a time and forward the request based on the URL supplied. So I think this behavior just kind of surprised me! But let's stick with this for now and if other people are confused for the same reason, we could revisit this. The new check you've added here should help a lot :)

Reply:
> I had actually thought that the default URL was always app.viam.com, and that if you want to connect/send requests to a different deployment you had to include that URL in every request

Ah that is correct! That base URL was not cached before this PR, and if you wanted to operate against a different deployment you had to include that URL as `--base-url` in every action you took on the CLI.

> And I had thought the CLI could have you logged into multiple sessions at a time and forward the request based on the URL supplied

Right, there was/continues to be no support for multiple sessions against different base URLs.

Reply:
> But let's stick with this for now and if other people are confused for the same reason, we could revisit this. The new check you've added here should help a lot :)

Sounds good to me! Thanks for the quick reviews 😄

Reply:
> That base URL was not cached before this PR, and if you wanted to operate against a different deployment you had to include that URL as --base-url in every action you took on the CLI.

Oh cool! Okay I think that behavior in particular was what made me think I didn't have to log out of a previous session before logging into a new one. This new error message should really help clear things up though! I even just did a quick test of the current behavior and if I run `viam auth --base-url=app.viam.dev` while logged into `app.viam.com`, it says `Already authenticated as "devin.hilly@viam.com" expires at 2023-08-30 15:57:41.731296 -0400 EDT m=+3612.657035834`, which is pretty misleading. So really excited for this improvement! 😄

Reply:
For sure; you used to be able to login with a new base-url without error; it would just say "already authenticated". Now it should say something like `cached base URL for this session is "https://app.viam.com:443", please logout and login again to use provided base URL "app.viam.dev"`.